src/poly/commitment.rs

```diff
@@ -154,6 +217,95 @@ impl<C: CurveAffine> Params<C> {
     }
 }
 
+/// A guard returned by the verifier
+#[derive(Debug)]
+pub struct Guard<'a, C: CurveAffine> {
+    /// Vector of random generators
+    pub g: Vec<C>,
+
+    /// Random generator
+    pub h: C,
+
+    /// Negation of z1 value in the OpeningProof
+    pub neg_z1: C::Scalar,
+
+    /// Params that were used by the verifier
+    pub params: &'a Params<C>,
+
+    /// Scalars produced by the verifier for multiscalar multiplication
+    pub scalars: Vec<C::Scalar>,
+
+    /// Points produced by the verifier for multiscalar multiplication
+    pub bases: Vec<C>,
+}
+
+impl<'a, C: CurveAffine> Guard<'a, C> {
+    /// Lets caller supply the challenges and obtain an MSM with updated
+    /// scalars and points.
+    pub fn use_challenges(
+        &mut self,
+        challenges_sq_packed: Vec<Challenge>,
+    ) -> Result<MSM<C>, Error> {
+        // - [z1] G
+        let mut allinv = C::Scalar::one();
+        let mut challenges_sq = Vec::with_capacity(self.params.k as usize);
+
+        for challenge_sq_packed in challenges_sq_packed {
+            let challenge_sq: C::Scalar = get_challenge_scalar(challenge_sq_packed);
+            challenges_sq.push(challenge_sq);
+
+            let challenge = challenge_sq.deterministic_sqrt();
+            if challenge.is_none() {
+                // We didn't sample a square.
+                return Err(Error::OpeningError);
+            }
+            let challenge = challenge.unwrap();
+
+            let challenge_inv = challenge.invert();
+            if bool::from(challenge_inv.is_none()) {
+                // We sampled zero for some reason, unlikely to happen by
+                // chance.
+                return Err(Error::OpeningError);
+            }
+            let challenge_inv = challenge_inv.unwrap();
+            allinv *= &challenge_inv;
+        }
+
+        self.bases.extend(&self.g);
+        let mut s = compute_s(&challenges_sq, allinv);
+        // TODO: parallelize
+        for s in &mut s {
+            *s *= &self.neg_z1;
+        }
+        self.scalars.extend(s);
+
+        Ok(MSM {
```

Comment:
This returns a new `MSM`, but what we want is to operate on (and return) the MSM located in `self`.